server/src/main/java/org/elasticsearch/cluster/ClusterStatePublicationEvent.java

```diff
@@ -35,4 +51,110 @@ public ClusterState getOldState() {
+     */
+    public long getPublicationContextConstructionElapsedMillis() {
+        assert publicationContextConstructionElapsedMillis != NOT_SET;
+        return publicationContextConstructionElapsedMillis;
+    }
+
+    /**
+     * @return how long in milliseconds it took to commit the publication, i.e. the elapsed time from the publication start until the master
+     *         receives publish responses from a majority of master nodes indicating that the state has been received and persisted there.
+     */
+    public long getPublicationCommitElapsedMillis() {
+        assert publicationCommitElapsedMillis != NOT_SET;
+        return publicationCommitElapsedMillis;
+    }
+
+    /**
+     * @return how long in milliseconds it took to complete the publication, i.e. the elapsed time from the publication start until all
+     *         nodes except the master have applied the cluster state.
+     */
+    public long getPublicationCompletionElapsedMillis() {
+        assert publicationCompletionElapsedMillis != NOT_SET;
+        return publicationCompletionElapsedMillis;
+    }
+
+    /**
+     * @return how long in milliseconds it took for the master to apply the cluster state, which happens after publication completion.
+     */
+    public long getMasterApplyElapsedMillis() {
+        assert masterApplyElapsedMillis != NOT_SET;
+        return masterApplyElapsedMillis;
+    }
+
+    /**
+     * @return how long in milliseconds it took to construct the publication context, which includes computing a cluster state diff and
+     *         serializing the cluster states for future transmission, or zero if not set.
+     */
+    public long maybeGetPublicationContextConstructionElapsedMillis() {
+        return ifSet(publicationContextConstructionElapsedMillis);
+    }
+
+    /**
+     * @return how long in milliseconds it took to commit the publication, i.e. the elapsed time from the publication start until the master
+     *         receives publish responses from a majority of master nodes indicating that the state has been received and persisted there,
+     *         or zero if not set.
+     */
+    public long maybeGetPublicationCommitElapsedMillis() {
+        return ifSet(publicationCommitElapsedMillis);
+    }
+
+    /**
+     * @return how long in milliseconds it took to complete the publication, i.e. the elapsed time from the publication start until all
+     *         nodes except the master have applied the cluster state, or zero if not set.
+     */
+    public long maybeGetPublicationCompletionElapsedMillis() {
+        return ifSet(publicationCompletionElapsedMillis);
+    }
+
+    /**
+     * @return how long in milliseconds it took for the master to apply the cluster state, which happens after publication completion, or
+     *         zero if not set.
+     */
+    public long maybeGetMasterApplyElapsedMillis() {
+        return ifSet(masterApplyElapsedMillis);
+    }
+
+    private static long ifSet(long millis) {
```

Comment:
Used by the `maybeGet` functions for failed-publication stats since the failure could happen at any stage.